File: backend/app/services/liver_segmentation_service.rb
# Сервис для выполнения сегментации печени на КТ-сканах
# Отвечает за:
# - Создание задачи сегментации
# - Оркестрацию процесса сегментации (предобработка, инференс, постобработка)
# - Интеграцию с Python сервисом нейросети
# - Сохранение результатов с метриками качества
# - Управление статусами задачи и КТ-скана
class LiverSegmentationService < ApplicationService
  attr_reader :ct_scan, :task, :error

  # Инициализация сервиса
  # Параметры:
  #   - ct_scan: Объект CtScan для сегментации
  def initialize(ct_scan)
    @ct_scan = ct_scan
    @error = nil
  end

  # Основной метод выполнения сегментации
  # Возвращает: OpenStruct с результатом (success?, result, error)
  def call
    # Проверка наличия КТ-скана
    return failure('CT scan not found') unless ct_scan
    # Проверка, не обработан ли уже КТ-скан
    return failure('CT scan already processed') if ct_scan.processed?

    # Выполнение в транзакции для обеспечения целостности данных
    ActiveRecord::Base.transaction do
      @task = create_segmentation_task  # Создание задачи
      process_segmentation  # Обработка сегментации
      success(task)  # Возврат успешного результата
    end
  rescue StandardError => e
    # Обработка ошибок
    Rails.logger.error("Segmentation failed: #{e.message}")
    Rails.logger.error(e.backtrace.join("\n"))
    @task&.mark_as_failed!(e.message)  # Пометить задачу как неудачную
    failure(e.message)
  end

  private

  # Создание задачи сегментации
  def create_segmentation_task
    ct_scan.segmentation_tasks.create!(status: 'pending')
  end

  # Основной процесс сегментации
  # Выполняет:
  # 1. Подготовку данных для нейросети
  # 2. Запуск инференса (вызов Python сервиса)
  # 3. Создание записи результата с метриками
  # 4. Обновление статусов
  def process_segmentation
    # Пометить задачу и КТ-скан как обрабатывающиеся
    task.mark_as_processing!
    ct_scan.update!(status: 'processing')

    # Подготовка входных данных для нейросети
    input_data = prepare_input_data
    
    # Запуск инференса нейросети (интеграция с Python сервисом)
    inference_result = run_inference(input_data)
    
    # Создание записи результата с метриками
    create_result(inference_result)
    
    # Обновление статусов на завершенные
    task.mark_as_completed!
    ct_scan.update!(status: 'completed')
  end

  # Подготовка входных данных для нейросети
  # Возвращает: Hash с данными для Python сервиса
  def prepare_input_data
    {
      ct_scan_id: ct_scan.id,
      patient_id: ct_scan.patient_id,
      dicom_path: ct_scan.dicom_file.attached? ? active_storage_path(ct_scan.dicom_file) : nil,  # Путь к DICOM файлу
      slice_count: ct_scan.slice_count,  # Количество срезов
      modality: ct_scan.modality  # Модальность (CT, MR и т.д.)
    }
  end

  # Запуск инференса нейросети
  # В продакшене здесь должен быть вызов Python сервиса
  # Сейчас возвращает мок-данные для тестирования
  # Параметры:
  #   - input_data: Hash с входными данными
  # Возвращает: Hash с результатами инференса (маска, контуры, метрики)
  def run_inference(input_data)
    # TODO: В продакшене здесь должен быть вызов Python сервиса:
    # PythonInferenceService.new.segment_liver(input_data)
    
    Rails.logger.info("Running inference for CT scan #{ct_scan.id}")
    
    # Симуляция инференса с мок-данными (заменить на реальный вызов Python сервиса)
    {
      mask_data: generate_mock_mask,  # Данные маски сегментации
      contours: generate_mock_contours,  # Данные контуров
      metrics: calculate_mock_metrics,  # Метрики качества
      inference_time_ms: rand(5000..15000)  # Время выполнения инференса (мс)
    }
  end

  # Создание записи результата сегментации
  # Параметры:
  #   - inference_result: Hash с результатами инференса
  # Возвращает: Объект SegmentationResult
  def create_result(inference_result)
    # Создание результата с метриками
    result = task.build_segmentation_result(
      dice_coefficient: inference_result[:metrics][:dice],  # Коэффициент Соренсена-Дайса
      iou_score: inference_result[:metrics][:iou],  # Intersection over Union
      volume_ml: inference_result[:metrics][:volume_ml],  # Объем печени в миллилитрах
      metrics: inference_result[:metrics],  # Все метрики в формате JSON
      contours: inference_result[:contours]  # Данные контуров в формате JSON
    )
    
    # Прикрепление файла маски, если он есть
    if inference_result[:mask_data]
      result.mask_file = inference_result[:mask_data][:path]
    end
    
    result.save!
    result
  end

  # Генерация мок-данных маски (для тестирования)
  # В продакшене маска приходит из Python сервиса
  def generate_mock_mask
    {
      format: 'nifti',  # Формат файла маски
      path: 'tmp/masks/mock_mask.nii.gz',  # Путь к файлу
      dimensions: [512, 512, ct_scan.slice_count || 100]  # Размеры маски
    }
  end

  # Генерация мок-данных контуров (для тестирования)
  # В продакшене контуры извлекаются из маски Python сервисом
  def generate_mock_contours
    {
      format: 'json',
      slices: (0...([ct_scan.slice_count, 10].min)).map do |i|
        {
          slice_index: i,
          contour_points: generate_random_contour_points  # Точки контура
        }
      end
    }
  end

  # Генерация случайных точек контура (для тестирования)
  def generate_random_contour_points
    # Генерация простого кругового контура для демонстрации
    center_x = 256
    center_y = 256
    radius = 80 + rand(40)
    
    (0...36).map do |i|
      angle = (i * 10) * Math::PI / 180
      {
        x: (center_x + radius * Math.cos(angle)).round(2),
        y: (center_y + radius * Math.sin(angle)).round(2)
      }
    end
  end

  # Расчет мок-метрик (для тестирования)
  # В продакшене метрики рассчитываются Python сервисом
  # Генерирует реалистичные значения, соответствующие клиническим стандартам (>= 0.90)
  def calculate_mock_metrics
    {
      dice: (0.90 + rand * 0.07).round(4),  # 0.90 - 0.97 (отличное качество)
      iou: (0.89 + rand * 0.08).round(4),    # 0.89 - 0.97
      volume_ml: (1200.0 + rand * 400.0).round(2), # 1200 - 1600 мл (типичный объем печени)
      pixel_accuracy: (0.95 + rand * 0.04).round(4),  # Точность пикселей
      sensitivity: (0.92 + rand * 0.06).round(4),  # Чувствительность
      specificity: (0.96 + rand * 0.03).round(4)  # Специфичность
    }
  end

  # Получение пути к файлу в Active Storage
  # Параметры:
  #   - attachment: Объект Active Storage attachment
  # Возвращает: Путь к файлу на диске
  def active_storage_path(attachment)
    return nil unless attachment.attached?
    
    if attachment.service.respond_to?(:path_for)
      attachment.service.path_for(attachment.key)
    else
      # Для локального хранилища
      Rails.root.join('storage', attachment.key).to_s
    end
  end

  # Возврат успешного результата
  def success(result)
    OpenStruct.new(success?: true, result: result, error: nil)
  end

  # Возврат результата с ошибкой
  def failure(message)
    @error = message
    OpenStruct.new(success?: false, result: nil, error: message)
  end
end

File: backend/app/services/liver_segmentation_service.py
"""
Service for liver segmentation
"""
import json
import math
import random
from typing import Dict, Any, Optional
from sqlalchemy.orm import Session

from app.models.ct_scan import CtScan, CtScanStatus
from app.models.segmentation_task import SegmentationTask, SegmentationTaskStatus
from app.models.segmentation_result import SegmentationResult
from app.services.application_service import ApplicationService, ServiceResult
from app.config import get_settings

settings = get_settings()


class LiverSegmentationService(ApplicationService):
    """Service for liver segmentation"""
    
    def __init__(self, ct_scan: CtScan, db: Session = None):
        self.ct_scan = ct_scan
        self.db = db
        self.task: Optional[SegmentationTask] = None
        self.error = None
    
    def execute(self) -> ServiceResult:
        """Execute segmentation"""
        if not self.ct_scan:
            return self.failure("CT scan not found")
        
        if self.ct_scan.is_processed():
            return self.failure("CT scan already processed")
        
        try:
            # Create task
            self.task = self._create_segmentation_task()
            
            # Process segmentation
            self._process_segmentation()
            
            return self.success(self.task)
        except Exception as e:
            if self.task:
                self.task.mark_as_failed(str(e))
                self.db.commit()
            return self.failure(str(e))
    
    def _create_segmentation_task(self) -> SegmentationTask:
        """Create segmentation task"""
        task = SegmentationTask(
            ct_scan_id=self.ct_scan.id,
            status=SegmentationTaskStatus.PENDING
        )
        self.db.add(task)
        self.db.commit()
        self.db.refresh(task)
        return task
    
    def _process_segmentation(self):
        """Process segmentation"""
        # Mark as processing
        self.task.mark_as_processing()
        self.ct_scan.status = CtScanStatus.PROCESSING
        self.db.commit()
        
        # Prepare input data
        input_data = self._prepare_input_data()
        
        # Run inference
        inference_result = self._run_inference(input_data)
        
        # Create result
        self._create_result(inference_result)
        
        # Mark as completed
        self.task.mark_as_completed(inference_result.get("inference_time_ms"))
        self.ct_scan.status = CtScanStatus.COMPLETED
        self.db.commit()
    
    def _prepare_input_data(self) -> Dict[str, Any]:
        """Prepare input data for neural network"""
        return {
            "ct_scan_id": self.ct_scan.id,
            "patient_id": self.ct_scan.patient_id,
            "dicom_path": self.ct_scan.dicom_file if hasattr(self.ct_scan, 'dicom_file') else None,
            "slice_count": self.ct_scan.slice_count,
            "modality": self.ct_scan.modality
        }
    
    def _run_inference(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Run neural network inference"""
        # TODO: In production, call Python neural network service
        # For now, return mock data
        
        return {
            "mask_data": self._generate_mock_mask(),
            "contours": self._generate_mock_contours(),
            "metrics": self._calculate_mock_metrics(),
            "inference_time_ms": random.randint(5000, 15000)
        }
    
    def _create_result(self, inference_result: Dict[str, Any]):
        """Create segmentation result"""
        metrics = inference_result["metrics"]
        
        result = SegmentationResult(
            segmentation_task_id=self.task.id,
            dice_coefficient=metrics["dice"],
            iou_score=metrics["iou"],
            volume_ml=metrics["volume_ml"],
            metrics=metrics,
            contours=inference_result["contours"]
        )
        
        if inference_result.get("mask_data"):
            result.mask_file = inference_result["mask_data"].get("path")
        
        self.db.add(result)
        self.db.commit()
    
    def _generate_mock_mask(self) -> Dict[str, Any]:
        """Generate mock mask data"""
        return {
            "format": "nifti",
            "path": f"tmp/masks/mock_mask_{self.ct_scan.id}.nii.gz",
            "dimensions": [512, 512, self.ct_scan.slice_count or 100]
        }
    
    def _generate_mock_contours(self) -> Dict[str, Any]:
        """Generate mock contour data"""
        slice_count = min(self.ct_scan.slice_count or 100, 10)
        
        return {
            "format": "json",
            "slices": [
                {
                    "slice_index": i,
                    "contour_points": self._generate_random_contour_points()
                }
                for i in range(slice_count)
            ]
        }
    
    def _generate_random_contour_points(self) -> list:
        """Generate random contour points"""
        center_x = 256
        center_y = 256
        radius = 80 + random.randint(0, 40)
        
        points = []
        for i in range(36):
            angle = (i * 10) * math.pi / 180
            points.append({
                "x": round(center_x + radius * math.cos(angle), 2),
                "y": round(center_y + radius * math.sin(angle), 2)
            })
        
        return points
    
    def _calculate_mock_metrics(self) -> Dict[str, float]:
        """Calculate mock metrics"""
        return {
            "dice": round(0.90 + random.random() * 0.07, 4),
            "iou": round(0.89 + random.random() * 0.08, 4),
            "volume_ml": round(1200.0 + random.random() * 400.0, 2),
            "pixel_accuracy": round(0.95 + random.random() * 0.04, 4),
            "sensitivity": round(0.92 + random.random() * 0.06, 4),
            "specificity": round(0.96 + random.random() * 0.03, 4)
        }

